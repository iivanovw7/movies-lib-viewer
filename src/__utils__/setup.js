/**
 * Module contains tests setup
 * @module __utils__/setup
 * @author Igor Ivanov
 */
import 'core-js/stable';
import 'regenerator-runtime/runtime';
import Enzyme, { configure } from 'enzyme';
import jsdom from 'jsdom';

// eslint-disable-next-line import/no-extraneous-dependencies
const Adapter = require('enzyme-adapter-react-16');

Enzyme.configure({ adapter: new Adapter() });

const dom = new jsdom.JSDOM('<!DOCTYPE html><html><body></body></html>', {
  url: 'http://localhost:4428',
  pretendToBeVisual: true,
});
const win = dom.window;

global.document = win.document;
global.window = win;
global.localStorage = win.localStorage;
global.requestAnimationFrame = win.requestAnimationFrame;
global.cancelAnimationFrame = win.cancelAnimationFrame;

// eslint-disable-next-line no-shadow
global.window.IntersectionObserver = class IntersectionObserver {
  // eslint-disable-next-line no-useless-constructor,require-jsdoc,no-empty-function
  constructor() {}

  // eslint-disable-next-line class-methods-use-this,require-jsdoc
  observe(cb) {
    return cb;
  }

  // eslint-disable-next-line class-methods-use-this,require-jsdoc
  unobserve() {
    return null;
  }
};

global.window.IntersectionObserverEntry = {};
global.window.IntersectionObserverEntry.prototype = {
  intersectionRatio: {},
};

configure({ adapter: new Adapter() });
