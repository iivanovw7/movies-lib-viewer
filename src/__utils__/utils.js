/**
 * Module contains test utils
 * @module __utils__/utils
 * @author Igor Ivanov
 */
import chalk from 'chalk';
/* eslint-disable */
export const testName = (title, description) => `[${chalk.yellow(title)}]: ${description}`;
/* eslint-enable */
