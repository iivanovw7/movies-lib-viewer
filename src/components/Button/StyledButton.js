/**
 * Module contains styled button for switch component
 * @module components/Switch/StyledButton
 * @author Igor Ivanov
 */
import styled from 'styled-components';

const StyledButton = styled.button`
  background: transparent;
  border: none;
  ${(props) => props.styling};
  opacity: ${(props) => (props.hide ? '0' : '1')};
  outline: none;
  transition: opacity 0.5s ease-in-out;

  &:hover {
    cursor: pointer;
    user-select: none;
  }
`;

export default StyledButton;
