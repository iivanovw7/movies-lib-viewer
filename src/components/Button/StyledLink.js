/**
 * Module contains styled link component
 * @module components/Button/StyledLink
 * @author Igor Ivanov
 */
import styled from 'styled-components';

const StyledLink = styled.a`
  border: none;
  ${(props) => props.styling}
  outline: none;

  &:hover {
    cursor: pointer;
    user-select: none;
  }
`;

export default StyledLink;
