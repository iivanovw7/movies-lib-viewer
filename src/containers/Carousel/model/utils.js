/**
 * Module contains carousel page utils
 * @module containers/Carousel/model/utils
 */

import setting from '../../../config/data';

/**
 * Calculates previous position.
 * @param {number} length - slides length.
 * @param {number} current - current slide.
 * @param {number} [skip = 0] - slide to be skipped.
 * @return {number} previous slide position.
 */
export function previous(length, current, skip = 0) {
  return Number((current - 1 - skip + length) % length);
}

/**
 * Calculates next position.
 * @param {number} length - slides length.
 * @param {number} current - current slide.
 * @param {number} [skip = 0] - slide to be skipped.
 * @return {number} next slide position.
 */
export function next(length, current, skip = 0) {
  return Number((current + 1 + skip) % length);
}

/**
 * Returns carousel element threshold.
 * @param {Object} target - target element.
 * @return {number} target threshold
 */
export function threshold(target) {
  const width = target.clientWidth;
  return width / 3;
}

/**
 * Carousel swipe event handler.
 * @param {Object} eventData data.
 * @param {number} length - slides length.
 * @param {number} dir - swipe direction.
 * @return {null | number} swipe length or null.
 */
export function swiped(eventData, length, dir = -1) {
  const t = threshold(eventData.event.target);
  const d = dir * eventData.deltaX;

  return d >= t ? length : null;
}

const basePosterUrl = (size) => `${setting.net.tmdbImageUrl}/t/p/w${size}/`;

export const posterUrl = (slide, size) => `${basePosterUrl(size)}${slide.poster_path}`;
