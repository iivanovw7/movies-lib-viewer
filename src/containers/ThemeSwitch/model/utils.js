/**
 * Module contains utils used in Theme Switch component
 * @module containers/ThemeSwitch/model/utils
 * @author Igor Ivanov
 */
import { lt, indexOf, inc, always, nth, ifElse, head, pipe, dec, __, length, keys, either } from 'ramda';

import { THEME_RESOURCES, DEFAULT_THEME } from './constants';

/**
 *  Gets string representations of keys out of theme object
 *  @return {Array.<string>} theme names list
 */
const themes = keys(THEME_RESOURCES);
/**
 * Gets index of current locale.
 * @type {f1}
 */
const themeIndex = indexOf(__, themes);
/**
 * Gets length of theme list.
 */
const themesLength = length(themes);
/**
 * Verifies if there is next theme available.
 * @type {f1}
 */
const hasNextTheme = lt(__, dec(themesLength));
/**
 * Returns firs theme, in case if no further themes available.
 */
const firstTheme = head(themes);

/**
 * Function selects next theme name.
 * Returns theme with next index if there is available one or first theme otherwise.
 * @type {Function|*}
 */
const newThemeKey = pipe(themeIndex, ifElse(hasNextTheme, pipe(inc, nth(__, themes)), always(firstTheme)));

/**
 * Returns either a new theme key or default key.
 * @type {Function|*}
 */
const nextThemeKey = either(newThemeKey, DEFAULT_THEME);

export default nextThemeKey;
