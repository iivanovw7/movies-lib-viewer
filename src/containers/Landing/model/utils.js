/**
 * Module contains landing page utils
 * @module containers/Landing/model/utils
 */
import { filter, pipe, prop } from 'ramda';

import { isString } from '../../../utils/helpers';

// eslint-disable-next-line import/prefer-default-export
export const withPosters = (list) => filter(pipe(prop('poster_path'), isString))(list);
