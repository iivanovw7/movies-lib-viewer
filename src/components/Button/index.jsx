/**
 * Module contains button component
 * @module components/Button
 * @author Igor Ivanov
 */
import * as PropTypes from 'prop-types';
import * as R from 'ramda';
import React from 'react';

import { isUndefined } from '../../utils/helpers';

import StyledButton from './StyledButton';
import StyledLink from './StyledLink';

/**
 * Creates styled button component.
 * @name Components/Button
 * @method
 *
 * @param {Object.<module:components/Button~propTypes>} props
 *  contains component props
 *  @see {@link module:components/Button~propTypes}
 * @return {Node} React component with children.
 * @constructor
 */
const Button = (props) => {
  const { variant, children, styling, onClick, index } = props;

  /* eslint-disable react/jsx-props-no-spreading */
  const ButtonElement = R.ifElse(
    R.pipe(R.prop('onClick'), isUndefined),
    R.always(<StyledLink {...props} />),
    R.always(<StyledButton {...props} />),
  );
  /* eslint-enable react/jsx-props-no-spreading */

  return (
    <ButtonElement data-index={index} variant={variant} styling={styling} onClick={onClick}>
      {children}
    </ButtonElement>
  );
};

/**
 * @name propTypes
 * @type {Object}
 * @param {Object} props - React PropTypes
 * @property {string} [props.variant = 'primary'] - color variant string.
 * @property {string | Image} props.children - button children element or text.
 * @property {Array.<string>} [props.styling = []] - image additional styles.
 * @property {Function} props.onClick - click handler.
 * @property {string} props.href - target link.
 * @property {string} [props.target = '_blank'] - target prop.
 * @property {boolean} [props.hidden = false] - if element should be hidden.
 * @property {number} [props.index = 0] - index which is passed in `data-index` attribute.
 * @return {Array} React propTypes
 */
Button.propTypes = {
  variant: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.func, PropTypes.node])),
  ]),
  styling: PropTypes.array,
  onClick: PropTypes.func,
  href: PropTypes.string,
  target: PropTypes.string,
  hidden: PropTypes.bool,
  index: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

Button.defaultProps = {
  target: '_blank',
  hidden: false,
  index: 0,
};

export default Button;
