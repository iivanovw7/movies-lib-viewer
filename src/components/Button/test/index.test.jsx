/**
 * Module contains Button component test suite
 */
import { mount, shallow } from 'enzyme';
import React from 'react';
import { css } from 'styled-components';

import ChildMock from '../../../__mocks__/childMock';
import { noop } from '../../../__utils__/stub';
import { testName } from '../../../__utils__/utils';
import Button from '../index';

// prettier-ignore
describe(testName('Button', 'component test suite'), function ButtonTestSuite() {
  // eslint-disable-next-line require-jsdoc
  const Composition = (props, children, shouldMount = false) => {
    // eslint-disable-next-line react/jsx-props-no-spreading
    const node = <Button {...props}>{children}</Button>;

    return shouldMount ? mount(node) : shallow(node);
  };

  let undef;
  const baseProps = {
    variant: 'primary',
    onClick: noop,
    href: 'href',
    target: 'target',
    styling: [],
  };

  it('Should match snapshot', () => {
    const fakeProps = {
      ...baseProps,
    };
    const component = Composition(fakeProps, ChildMock.component, false);

    expect(component)
      .toMatchSnapshot();
  });

  it(testName('button', 'Should mount and check if onClick has been passed in props'), () => {
    const fakeProps = {
      ...baseProps,
    };
    const component = Composition(fakeProps, ChildMock.component, true);

    expect(component.find('button'))
      .toHaveLength(1);
  });
  // prettier-ignore
  it(testName('button: a', 'Should mount as link and check if no onClick has been passed in props'), () => {
    const fakeProps = {
      ...baseProps,
      onClick: undef,
    };
    const component = Composition(fakeProps, ChildMock.component, true);

    expect(component.find('a'))
      .toHaveLength(1);
  });
  // prettier-ignore
  it(testName('button', 'Should verify if all attributes have been passed to child component'), () => {
    const ButtonStyles = css`
      height: 1.2em;
    `;
    const fakeProps = {
      ...baseProps,
      onClick: undef,
      href: 'test_href',
      target: '_blank',
      styling: ButtonStyles,
    };
    const component = Composition(fakeProps, ChildMock.component, true);
    const child = component.find(`#${ChildMock.id}`);
    const button = component.find(Button);

    expect(child)
      .toHaveLength(1);

    expect(child.text())
      .toEqual(ChildMock.child);

    expect(button.prop('target'))
      .toEqual(fakeProps.target);

    expect(button.prop('href'))
      .toEqual(fakeProps.href);

    expect(button.prop('styling'))
      .toBe(ButtonStyles);
  });

  it(testName('button', 'Should simulate onClick event'), () => {
    const onClick = jest.fn();
    const fakeProps = {
      ...baseProps,
      onClick,
    };
    const component = Composition(fakeProps, ChildMock.child, true);

    component.find(Button)
      .at(0)
      .simulate('click');

    expect(onClick)
      .toHaveBeenCalled();
  });
});
