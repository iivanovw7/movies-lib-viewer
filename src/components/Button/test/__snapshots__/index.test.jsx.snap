// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[[33mButton[39m]: component test suite Should match snapshot 1`] = `
<f1
  data-index={0}
  onClick={[Function]}
  styling={Array []}
  variant="primary"
>
  <div
    id="child-mock-id"
  >
    child-mock-id
  </div>
</f1>
`;
